build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaModulePathPlugin.java

```diff
@@ -0,0 +1,194 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.internal;
+
+import org.elasticsearch.gradle.VersionProperties;
+import org.elasticsearch.gradle.util.GradleUtils;
+import org.gradle.api.Action;
+import org.gradle.api.Named;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+import org.gradle.api.artifacts.component.ProjectComponentIdentifier;
+import org.gradle.api.artifacts.result.ResolvedComponentResult;
+import org.gradle.api.artifacts.result.ResolvedDependencyResult;
+import org.gradle.api.attributes.LibraryElements;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.compile.JavaCompile;
+import org.gradle.process.CommandLineArgumentProvider;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * The Java Module Compile Path Plugin, i.e. --module-path, ---module-version
+ */
+public class ElasticsearchJavaModulePathPlugin implements Plugin<Project> {
+    @Override
+    public void apply(Project project) {
+        configureCompileModulePath(project);
+    }
+
+    // List of root tasks, by name, whose compileJava task should not use the module path. These are test related sources.
+    static final Set<String> EXCLUDES = Set.of(":test:framework", ":x-pack:plugin:eql:qa:common");
+
+    static void configureCompileModulePath(Project project) {
+        // first disable Gradle's builtin module path inference
+        project.getTasks()
+            .withType(JavaCompile.class)
+            .configureEach(compileTask -> compileTask.getModularity().getInferModulePath().set(false));
+
+        // test:framework has split pkgs with server, libs and more. do not use module path
+        var projName = project.toString();
+        if (EXCLUDES.stream().anyMatch(name -> projName.contains(name))) {
+            return;
+        }
+
+        var isModuleProject = hasModuleInfoDotJava(project);
+        var configurations = project.getConfigurations();
+        var compileClasspath = configurations.getByName(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME);
+
+        var moduleCompileClasspath = configurations.create("moduleCompileClasspath", it -> {
+            it.extendsFrom(compileClasspath);
+            it.setCanBeConsumed(false); // We don't want this configuration used by dependent projects
+            it.attributes(
```

Comment:
Lets add an explicit `it.setCanBeResolved(true);` to make the purpose of this configuration explicit